### PR TITLE
Introduce v2 runtime and generator

### PR DESCRIPTION
**Pull request by noxiouz**
Add a separate github.com/varlink/go/v2 module that can evolve independently from the v1 API while preserving the root module. The v2 design splits the implementation into explicit codec, wire, runtime, and generated API layers so framing, JSON handling, request semantics, and typed bindings can be optimized and reasoned about separately.

The internal codec now owns null-delimited framing, bounded frame reads, caller-owned buffer reuse, net.Buffers-based writes, and a JSON wire layer. It also supports both the default encoding/json backend and the experimental encoding/json/v2 backend behind GOEXPERIMENT=jsonv2. Frame and wire tests plus benchmarks cover small, medium, and large messages, fragmented reads, oversized frames, EOF behavior, and write-path allocation shape.

The v2 runtime adds explicit client and server APIs for unary calls, oneway requests, streamed replies, upgrades, and pipelined unary/oneway batches. Server registration is builder-based and immutable after Build, with separate handler types for unary, stream, and upgrade methods. The runtime includes built-in org.varlink.service support, service metadata, interface descriptions, typed remote errors, deadline-aware framed transport helpers, and certification-style coverage for the core varlink data shapes.

Add a lightweight v2 introspection package that extracts interface metadata and method modes from interface descriptions. Method mode is expressed through structured comments such as @mode stream, @mode oneway, and @mode upgrade, with unannotated methods defaulting to unary because plain varlink IDL does not encode transport mode. The certification IDL is annotated accordingly for streamed and oneway methods.

Add a v2 interface generator that reuses the existing IDL parser but emits github.com/varlink/go/v2/varlink bindings. Generated code now includes typed request/reply/error declarations, typed clients, typed batch helpers for unary and oneway calls, typed stream receivers, typed upgrade methods, mode-correct server handler adapters, interface descriptions, and remote-error dispatch. The legacy generator and parser also learn the official any type.

This commit intentionally keeps v1 code intact while creating a compileable v2 staging area. Remaining work is mostly around higher-level transport helpers, resolver integration, richer public schema APIs, external cross-language certification, and broader examples/CI coverage.

### CODE REVIEW

File: v2/varlink/server_runtime.go [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+		return nil, err
+	}
+	r.replied = true
+	r.upgraded = true
+	if err := r.transport.writeMessage(ctx, r.cfg.WriteTimeout, Message{
+		Parameters: params,
+	}); err != nil {
+		return nil, err
+	}
+	return r.conn, nil
```
Comment (cgwalters) [inline v2/varlink/server_runtime.go]: AI: *Important*: This returns the raw `r.conn` rather than `r.transport.detach()`. The client side does the right thing — `ConnClient.detach()` calls `transport.detach()` to wrap any bytes already buffered in the framed reader into a `bufferedConn` (runtime.go:93), and there's even a `TestConnClientUpgradePreservesBufferedBytes`. On the server, the transport's `codec.Wire` reads into a 4096-byte buffer, so a client that pipelines upgraded raw bytes immediately after the upgrade request can have those bytes over-read into the buffer and then silently dropped here. This should return `r.transport.detach()` for symmetry, and ideally get a server-side analogue of the buffered-bytes test.

File: v2/varlink/client_runtime.go [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+}
+
+func (s *clientStream) Recv(ctx context.Context, out any) error {
+	if s.done {
+		return io.EOF
+	}
+
+	msg, err := s.client.readReply(ctx, s.method)
+	if err != nil {
+		s.release()
```
Comment (cgwalters) [inline v2/varlink/client_runtime.go:219]: AI: *Important*: On the error paths in `Recv` (here and at the decode failure below), `s.release()` sets `s.client = nil` but `s.done` stays false. Only the success/last-reply path at the bottom sets `s.done = true`. So if a caller ignores the returned error and calls `Recv` again, the `if s.done` guard at the top passes and `s.client.readReply(...)` dereferences a nil `s.client` → panic. Set `s.done = true` whenever you release on error.

File: v2/internal/codec/json_backend_v2.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	jsonv1 "encoding/json"
+	json "encoding/json/v2"
+)
+
+type RawValue = jsonv1.RawMessage
+
+func marshalJSON(v any) ([]byte, error) {
+	return json.Marshal(v)
```
Comment (cgwalters) [inline v2/internal/codec/json_backend_v2.go]: AI: (low) The jsonv2 backend diverges from the std backend here: `encoding/json` (v1) HTML-escapes `<`,`>`,`&` and sorts map keys deterministically by default; `encoding/json/v2` does neither. The two backends still produce semantically-equivalent JSON, so this is harmless for varlink interop. But it does mean switching via GOEXPERIMENT=jsonv2 changes the exact wire bytes, and non-deterministic map ordering can break byte-level comparisons (golden tests, hashing/signing) and reproducibility. If the backends are meant to be interchangeable, consider matching v1 (HTML escaping + deterministic key ordering), or add a test asserting byte-for-byte equivalence so the divergence can't drift unnoticed.

File: v2/varlink/runtime.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+func setReadDeadline(conn io.ReadWriteCloser, ctx context.Context, timeout time.Duration) error {
+	r, ok := conn.(deadlineReader)
+	if !ok {
+		return nil
+	}
+	deadline, ok := resolveDeadline(ctx, timeout)
+	if !ok {
+		return r.SetReadDeadline(time.Time{})
```
Comment (cgwalters) [inline v2/varlink/runtime.go]: AI: A cancelable context with no deadline is effectively ignored for blocking I/O. `resolveDeadline` returns `ok=false` when there's no timeout and no ctx deadline, so this clears the read deadline entirely. If the caller passes a `context.WithCancel` ctx (no deadline) and then cancels it while a read is blocked on an idle socket, nothing unblocks the I/O — the goroutine/connection leak until the peer acts. Consider an `AfterFunc`/watchdog that forces a deadline (or closes the conn) on `ctx.Done()` so cancellation actually propagates to the socket. Same pattern applies to `setWriteDeadline`.

File: v2/varlink/server_runtime.go [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+		call := &unaryRequest{
+			baseRequest: baseRequest{
+				method:     msg.Method,
+				parameters: msg.Parameters,
+				transport:  transport,
+				cfg:        s.cfg,
+				oneway:     msg.Oneway,
+			},
+		}
+		err := handler.unary(ctx, call)
```
Comment (cgwalters) [inline v2/varlink/server_runtime.go]: AI: User-registered handlers (unary/stream/upgrade) are invoked with no `recover()`. A panic in any handler will unwind through the serve loop and can take down the connection-serving goroutine (or the process, depending on how `Serve` is driven). For an RPC server runtime it's usually worth recovering per-call, replying with an internal-error to the client, and closing just that connection.

File: v2/internal/codec/json_backend_std.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+func marshalParameters(v any) (*RawValue, error) {
+	if v == nil {
+		return nil, nil
+	}
+	switch raw := v.(type) {
+	case *RawValue:
+		return raw, nil
+	case RawValue:
+		copyRaw := raw
```
Comment (cgwalters) [inline v2/internal/codec/json_backend_std.go]: AI: `RawValue` is a `[]byte` alias, so `copyRaw := raw` copies only the slice header — the returned `*RawValue` still aliases the caller's backing array. If the caller reuses/mutates that buffer after this returns, the parameters embedded in the outgoing message change underneath you. If callers are expected to hand off ownership this may be fine, but it's a subtle aliasing contract worth either documenting or making safe with `copyRaw := append(RawValue(nil), raw...)`. (Same in json_backend_v2.go.)

File: v2/varlink/server_runtime.go [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+		if err != nil {
+			if call.replied || call.upgraded {
```
Comment (cgwalters) [inline v2/varlink/server_runtime.go]: AI: (low) Both arms of this `if` return the identical `false, err`, so the condition is dead — it can collapse to just `return false, err`. Same redundant branch appears in the `handlerKindStream` case below (lines 125-128).

File: v2/varlink/introspect/introspect.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+		return nil, fmt.Errorf("introspect: missing interface name")
+	}
+	return info, nil
+}
+
+func ParseMethodMode(doc string) (MethodMode, error) {
+	mode := ModeUnary
+	for _, line := range strings.Split(doc, "\n") {
+		line = strings.TrimSpace(line)
+		if !strings.HasPrefix(line, "@mode") {
```
Comment (cgwalters) [inline v2/varlink/introspect/introspect.go]: AI: (low) `HasPrefix(line, "@mode")` also matches lines like `@models foo`, which then fall into the `len(fields)!=2` / unknown-mode error paths rather than being treated as ordinary doc text. Matching on the first whitespace-separated token (`fields[0] == "@mode"`) would be more robust. Note this same `@mode` parsing logic is duplicated verbatim in the v2 generator's main.go — worth sharing one implementation.

File: v2/varlink/errors.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	"errors"
+
+	"github.com/varlink/go/v2/internal/codec"
+)
+
+var (
+	// ErrFrameTooLarge is returned when a peer sends a message larger than the configured limit.
+	ErrFrameTooLarge = codec.ErrFrameTooLarge
+	// ErrStreamClosed is returned when receiving from a closed stream.
+	ErrStreamClosed = errors.New("varlink: stream closed")
```
Comment (cgwalters) [inline v2/varlink/errors.go:13]: AI: (low) `ErrStreamClosed` appears unused — stream termination returns `io.EOF` instead. Either wire it up (e.g. return it from `clientStream.Close`/`Recv` after an explicit close) or drop it.

File: v2/internal/codec/frame.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+	return f.ReadFrameInto(r, nil)
+}
+
+// ReadFrameInto reads one null-delimited payload into dst and returns the payload slice.
+// The returned slice aliases dst when dst has enough capacity for the payload.
+func (f Framer) ReadFrameInto(r *bufio.Reader, dst []byte) ([]byte, error) {
+	maxFrameSize := resolveMaxFrameSize(f.MaxFrameSize)
+	frame := dst[:0]
+	if cap(frame) == 0 {
+		frame = make([]byte, 0, min(maxFrameSize, 4*1024))
```
Comment (cgwalters) [inline v2/internal/codec/frame.go]: AI: (low) The `4*1024` initial-scratch cap and the `4096` default bufio size (NewReader, and runtime.go:64 `NewWire(..., 4096, ...)`) are repeated magic numbers. A couple of named constants (e.g. `defaultScratchCap`, `defaultBufferSize`) would make the intent and the relationship between them clearer.

**Review by cgwalters — COMMENTED** [review]
Since I'm trying to contribute to this repo (ref https://github.com/varlink/go/pull/44 ) I thought I'd spend some tokens reviewing this PR.

Assisted-by: OpenCode (Claude Opus 4.8 + Gemini 3.1 Pro)

AI-generated review. Comments prefixed with `AI:` are unedited AI output. Verified locally: the v2 module builds, `go vet`, `go test`, `go test -race`, and `GOEXPERIMENT=jsonv2 go test ./internal/codec/...` all pass.

File: v2/varlink/server_runtime.go (hunk 1 above)
Comment (noxiouz) [reply v2/varlink/server_runtime.go]: Addressed in 57bc35d. Server-side upgrade now returns the detached transport so any bytes already buffered by the framed reader are preserved, and there is a server-side regression test for pipelined upgraded bytes.

File: v2/internal/codec/json_backend_v2.go (hunk 3 above)
Comment (noxiouz) [reply v2/internal/codec/json_backend_v2.go]: Addressed in 57bc35d. The jsonv2 backend now uses encoding/json DefaultOptionsV1 with the v2 encoder/decoder, and tests cover v1-compatible ordering, escaping, nil map/slice handling, and invalid UTF-8 replacement semantics.

File: v2/varlink/runtime.go (hunk 4 above)
Comment (noxiouz) [reply v2/varlink/runtime.go]: Addressed in 57bc35d. Read and write deadline setup now uses context.AfterFunc to force a deadline when a cancelable context is canceled, so blocking I/O wakes promptly. Ordered client I/O failures also close the client connection.

File: v2/varlink/server_runtime.go (hunk 5 above)
Comment (noxiouz) [reply v2/varlink/server_runtime.go]: Addressed in 57bc35d. Unary, stream, and upgrade handlers are now called through panic-recovering wrappers. If possible the server replies with org.varlink.service.InternalError, then closes the serving connection; the built-in service description declares that error.

File: v2/varlink/client_runtime.go (hunk 2 above)
Comment (noxiouz) [reply v2/varlink/client_runtime.go:219]: Addressed in 57bc35d. Stream receive error paths now mark the stream terminal before releasing the client, so repeated Recv calls cannot dereference a nil client. Explicit stream close also returns ErrStreamClosed on later Recv.

File: v2/internal/codec/json_backend_std.go (hunk 6 above)
Comment (noxiouz) [reply v2/internal/codec/json_backend_std.go]: Addressed in 57bc35d. MarshalParameters now copies RawValue inputs in both the std and jsonv2 backends, with tests proving caller buffer mutation no longer aliases into the outgoing parameters.
